Port WebExtension Permissions and Match Patterns to C++

ItsJamie9494 opened the pull request:
#### 6b4b64e0f2638de129ff167b5df8dfd88a9a9889
<pre>
Port WebExtension Permissions and Match Patterns to C++
<a href="https://webkit.org/b/281388">https://webkit.org/b/281388</a>

Reviewed by Timothy Hatcher.

This creates a new WebExtensionPermission class in Shared/Extensions to store the permission constants, instead of relying on the Cocoa API. This also ports the code in the manifest parser that deals with permissions and externally connectable match patterns.

* Source/WebKit/Shared/Extensions/WebExtensionPermission.cpp: Added.
(WebKit::WebExtensionPermission::activeTab):
(WebKit::WebExtensionPermission::alarms):
(WebKit::WebExtensionPermission::clipboardWrite):
(WebKit::WebExtensionPermission::contextMenus):
(WebKit::WebExtensionPermission::cookies):
(WebKit::WebExtensionPermission::declarativeNetRequest):
(WebKit::WebExtensionPermission::declarativeNetRequestFeedback):
(WebKit::WebExtensionPermission::declarativeNetRequestWithHostAccess):
(WebKit::WebExtensionPermission::menus):
(WebKit::WebExtensionPermission::nativeMessaging):
(WebKit::WebExtensionPermission::notifications):
(WebKit::WebExtensionPermission::scripting):
(WebKit::WebExtensionPermission::sidePanel):
(WebKit::WebExtensionPermission::storage):
(WebKit::WebExtensionPermission::tabs):
(WebKit::WebExtensionPermission::unlimitedStorage):
(WebKit::WebExtensionPermission::webNavigation):
(WebKit::WebExtensionPermission::webRequest):
* Source/WebKit/Shared/Extensions/WebExtensionPermission.h: Added.
* Source/WebKit/Sources.txt:
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
(WebKit::WebExtension::populateDeclarativeNetRequestPropertiesIfNeeded):
(WebKit::WebExtension::populateExternallyConnectableIfNeeded): Deleted.
(WebKit::WebExtension::hasSidePanel): Deleted.
(WebKit::WebExtension::supportedPermissions): Deleted.
(WebKit::WebExtension::requestedPermissions): Deleted.
(WebKit::WebExtension::optionalPermissions): Deleted.
(WebKit::WebExtension::requestedPermissionMatchPatterns): Deleted.
(WebKit::WebExtension::optionalPermissionMatchPatterns): Deleted.
(WebKit::WebExtension::externallyConnectableMatchPatterns): Deleted.
(WebKit::WebExtension::allRequestedMatchPatterns): Deleted.
(WebKit::WebExtension::populatePermissionsPropertiesIfNeeded): Deleted.
* Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
(WebKit::WebExtension::hasSidePanel):
(WebKit::WebExtension::supportedPermissions):
(WebKit::WebExtension::requestedPermissions):
(WebKit::WebExtension::optionalPermissions):
(WebKit::WebExtension::requestedPermissionMatchPatterns):
(WebKit::WebExtension::optionalPermissionMatchPatterns):
(WebKit::WebExtension::externallyConnectableMatchPatterns):
(WebKit::WebExtension::allRequestedMatchPatterns):
(WebKit::WebExtension::populateExternallyConnectableIfNeeded):
(WebKit::WebExtension::populatePermissionsPropertiesIfNeeded):
* Source/WebKit/UIProcess/Extensions/WebExtension.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/285084@main">https://commits.webkit.org/285084@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/03a5a615cf8e73a485cbffce2913e2c71eb4d991

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/71508 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/50921 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/24282 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/75619 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/22713 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/58721 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/22532 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/56474 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/14943 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/74574 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/46205 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/61596 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/36923 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/42868 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/21054 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/64771 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/19426 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/77338 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/15742 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/18601 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/64187 "Passed tests") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/15785 "Build is in progress. Recent messages:OS: Sonoma (14.5), Xcode: 15.4; Running apply-patch; Checked out pull request; Compiled WebKit (warnings)") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/61635 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/64183 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/12331 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/5970 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10958 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/46721 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/1500 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/47792 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/49076 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/47534 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -1137,6 +1147,229 @@ void WebExtension::populateContentScriptPropertiesIfNeeded()
+        WebExtensionPermission::declarativeNetRequestWithHostAccess(), WebExtensionPermission::menus(), WebExtensionPermission::nativeMessaging(), WebExtensionPermission::notifications(), WebExtensionPermission::scripting(),
+        WebExtensionPermission::storage(), WebExtensionPermission::tabs(), WebExtensionPermission::unlimitedStorage(), WebExtensionPermission::webNavigation(), WebExtensionPermission::webRequest(),
+#if ENABLE(WK_WEB_EXTENSIONS_SIDEBAR)
+        WebExtensionPermission::sidePanel(),
```
You don't declare this in `WebExtensionPermission`.

ItsJamie9494 replied:
Are you saying I don't declare the WK_WEB_EXTENSIONS_SIDEBAR, or the sidePanel permission?

xeenon replied:
`sidePanel()` permission was missing.

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -1137,6 +1147,229 @@ void WebExtension::populateContentScriptPropertiesIfNeeded()
+        result.add(matchPattern);
+
+    for (auto& injectedContent : m_staticInjectedContents) {
+        for (auto& matchPattern : injectedContent.includeMatchPatterns)
```
```suggestion
        for (Ref matchPattern : injectedContent.includeMatchPatterns)
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -1137,6 +1147,229 @@ void WebExtension::populateContentScriptPropertiesIfNeeded()
+    for (auto& matchPattern : m_permissionMatchPatterns)
+        result.add(matchPattern);
+
+    for (auto& matchPattern : m_externallyConnectableMatchPatterns)
```
```suggestion
    for (Ref matchPattern : m_externallyConnectableMatchPatterns)
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -1137,6 +1147,229 @@ void WebExtension::populateContentScriptPropertiesIfNeeded()
+
+    WebExtension::MatchPatternSet result;
+
+    for (auto& matchPattern : m_permissionMatchPatterns)
```
```suggestion
    for (Ref matchPattern : m_permissionMatchPatterns)
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -1137,6 +1147,229 @@ void WebExtension::populateContentScriptPropertiesIfNeeded()
+        for (auto matchPatternStringValue : *matchPatternStrings) {
+            auto matchPatternString = matchPatternStringValue->asString();
+
+            if (matchPatternString.isEmpty())
```
```suggestion
            auto matchPatternString = matchPatternStringValue->asString();
            if (matchPatternString.isEmpty())
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -1137,6 +1147,229 @@ void WebExtension::populateContentScriptPropertiesIfNeeded()
+            if (matchPatternString.isEmpty())
+                continue;
+
+            if (auto matchPattern = WebExtensionMatchPattern::getOrCreate(matchPatternString)) {
```
```suggestion
            if (RefPtr matchPattern = WebExtensionMatchPattern::getOrCreate(matchPatternString)) {
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -1137,6 +1147,229 @@ void WebExtension::populateContentScriptPropertiesIfNeeded()
+        }
+    }
+
+    m_externallyConnectableMatchPatterns = matchPatterns;
```
```suggestion
    m_externallyConnectableMatchPatterns = WTFMove(matchPatterns);
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -1137,6 +1147,229 @@ void WebExtension::populateContentScriptPropertiesIfNeeded()
+        });
+    }
+
+    if (shouldReportError || (matchPatterns.isEmpty() && (!extensionIDs || (extensionIDs && !extensionIDs->length()))))
```
Since the `WTFMove()` I added above, need to use `m_externallyConnectableMatchPatterns` here.
```suggestion
    if (shouldReportError || (m_externallyConnectableMatchPatterns.isEmpty() && (!extensionIDs || (extensionIDs && !extensionIDs->length()))))
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -1137,6 +1147,229 @@ void WebExtension::populateContentScriptPropertiesIfNeeded()
+                continue;
+
+            if (findMatchPatternsInPermissions) {
+                if (auto matchPattern = WebExtensionMatchPattern::getOrCreate(permission)) {
```
```suggestion
                if (RefPtr matchPattern = WebExtensionMatchPattern::getOrCreate(permission)) {
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -1137,6 +1147,229 @@ void WebExtension::populateContentScriptPropertiesIfNeeded()
+                if (permission.isEmpty())
+                    continue;
+
+                if (auto matchPattern = WebExtensionMatchPattern::getOrCreate(permission)) {
```
```suggestion
                if (RefPtr matchPattern = WebExtensionMatchPattern::getOrCreate(permission)) {
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -1137,6 +1147,229 @@ void WebExtension::populateContentScriptPropertiesIfNeeded()
+                continue;
+
+            if (findMatchPatternsInPermissions) {
+                if (auto matchPattern = WebExtensionMatchPattern::getOrCreate(permission)) {
```
```suggestion
                if (RefPtr matchPattern = WebExtensionMatchPattern::getOrCreate(permission)) {
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -1137,6 +1147,229 @@ void WebExtension::populateContentScriptPropertiesIfNeeded()
+                if (permission.isEmpty())
+                    continue;
+
+                if (auto matchPattern = WebExtensionMatchPattern::getOrCreate(permission)) {
```
```suggestion
                if (RefPtr matchPattern = WebExtensionMatchPattern::getOrCreate(permission)) {
```

ItsJamie9494 committed:
Port WebExtension Permissions and Match Patterns to C++
https://webkit.org/b/281388

Reviewed by Timothy Hatcher.

This creates a new WebExtensionPermission class in Shared/Extensions to store the permission constants, instead of relying on the Cocoa API. This also ports the code in the manifest parser that deals with permissions and externally connectable match patterns.

* Source/WebKit/Shared/Extensions/WebExtensionPermission.cpp: Added.
(WebKit::WebExtensionPermission::activeTab):
(WebKit::WebExtensionPermission::alarms):
(WebKit::WebExtensionPermission::clipboardWrite):
(WebKit::WebExtensionPermission::contextMenus):
(WebKit::WebExtensionPermission::cookies):
(WebKit::WebExtensionPermission::declarativeNetRequest):
(WebKit::WebExtensionPermission::declarativeNetRequestFeedback):
(WebKit::WebExtensionPermission::declarativeNetRequestWithHostAccess):
(WebKit::WebExtensionPermission::menus):
(WebKit::WebExtensionPermission::nativeMessaging):
(WebKit::WebExtensionPermission::notifications):
(WebKit::WebExtensionPermission::scripting):
(WebKit::WebExtensionPermission::sidePanel):
(WebKit::WebExtensionPermission::storage):
(WebKit::WebExtensionPermission::tabs):
(WebKit::WebExtensionPermission::unlimitedStorage):
(WebKit::WebExtensionPermission::webNavigation):
(WebKit::WebExtensionPermission::webRequest):
* Source/WebKit/Shared/Extensions/WebExtensionPermission.h: Added.
* Source/WebKit/Sources.txt:
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
(WebKit::WebExtension::populateDeclarativeNetRequestPropertiesIfNeeded):
(WebKit::WebExtension::populateExternallyConnectableIfNeeded): Deleted.
(WebKit::WebExtension::hasSidePanel): Deleted.
(WebKit::WebExtension::supportedPermissions): Deleted.
(WebKit::WebExtension::requestedPermissions): Deleted.
(WebKit::WebExtension::optionalPermissions): Deleted.
(WebKit::WebExtension::requestedPermissionMatchPatterns): Deleted.
(WebKit::WebExtension::optionalPermissionMatchPatterns): Deleted.
(WebKit::WebExtension::externallyConnectableMatchPatterns): Deleted.
(WebKit::WebExtension::allRequestedMatchPatterns): Deleted.
(WebKit::WebExtension::populatePermissionsPropertiesIfNeeded): Deleted.
* Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
(WebKit::WebExtension::hasSidePanel):
(WebKit::WebExtension::supportedPermissions):
(WebKit::WebExtension::requestedPermissions):
(WebKit::WebExtension::optionalPermissions):
(WebKit::WebExtension::requestedPermissionMatchPatterns):
(WebKit::WebExtension::optionalPermissionMatchPatterns):
(WebKit::WebExtension::externallyConnectableMatchPatterns):
(WebKit::WebExtension::allRequestedMatchPatterns):
(WebKit::WebExtension::populateExternallyConnectableIfNeeded):
(WebKit::WebExtension::populatePermissionsPropertiesIfNeeded):
* Source/WebKit/UIProcess/Extensions/WebExtension.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/285084@main
